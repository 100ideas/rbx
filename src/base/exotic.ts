import React from "react";

export interface ForwardRefAsExoticComponent<
  TOwnProps,
  TDefaultComponent extends React.ReactType<any>
>
  extends Pick<
    React.ForwardRefExoticComponent<any>,
    keyof React.ForwardRefExoticComponent<any>
  > {
  <TAsComponent extends React.ReactType<any>>(
    props: Prefer<
      React.PropsWithoutRef<TOwnProps & { as: TAsComponent }>,
      React.ComponentPropsWithRef<TAsComponent>
    >,
  ): React.ReactElement<any> | null;
  (
    props: Prefer<
      React.PropsWithoutRef<TOwnProps & { as?: never }>,
      React.ComponentPropsWithRef<TDefaultComponent>
    >,
  ): React.ReactElement<any> | null;
}

export function forwardRefAs<
  TOwnProps,
  TDefaultElement extends React.ReactType<any>
>(
  factory: React.RefForwardingComponent<
    any,
    TOwnProps & { as?: React.ReactType<any> }
  >,
  defaultProps: Partial<
    Prefer<
      React.PropsWithoutRef<TOwnProps & { as: TDefaultElement }> &
        React.RefAttributes<any>,
      React.ComponentPropsWithoutRef<TDefaultElement>
    >
  >,
) {
  const forward = React.forwardRef(factory);
  forward.defaultProps = defaultProps;
  return forward as ForwardRefAsExoticComponent<TOwnProps, TDefaultElement>;
}
