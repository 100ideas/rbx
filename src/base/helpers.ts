import classNames from "classnames";
import * as PropTypes from "prop-types";

import { Omit, Prefer } from "../types";
import { tuple } from "../utils";

/**
 * Float
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const FLOAT_PULLED_ALIGNMENTS = tuple("left", "right");
export type FloatPulledAlignments = (typeof FLOAT_PULLED_ALIGNMENTS)[number];

export type FloatHelpersProps = Partial<{
  /** Fixes an element's floating children */
  clearfix: boolean;
  /** Moves an element to the left or right */
  pull: FloatPulledAlignments;
}>;

export const floatHelpersPropTypes = {
  clearfix: PropTypes.bool,
  pull: PropTypes.oneOf(FLOAT_PULLED_ALIGNMENTS),
};

export const transformFloatHelpers: TransformFunc<FloatHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    floatHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clearfix, pull, ...rest } = props;

  return {
    className: classNames(
      {
        "is-clearfix": clearfix,
        [`is-pulled-${pull}`]: pull,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Overflow
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverflowHelpersProps = Partial<{
  /** Adds overflow hidden */
  clipped: boolean;
}>;

export const overflowHelpersPropTypes = {
  clipped: PropTypes.bool,
};

export const transformOverflowHelpers: TransformFunc<OverflowHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overflowHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clipped, ...rest } = props;

  return {
    className: classNames({ "is-clipped": clipped }, className),
    ...rest,
  };
};

/**
 * Overlay
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverlayHelpersProps = Partial<{
  /** Completely covers the first positioned parent */
  overlay: boolean;
}>;

export const overlayHelpersPropTypes = {
  overlay: PropTypes.bool,
};

export const transformOverlayHelpers: TransformFunc<OverlayHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overlayHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, overlay, ...rest } = props;

  return {
    className: classNames({ "is-overlay": overlay }, className),
    ...rest,
  };
};

/**
 * Typography
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const COLORS = tuple(
  "primary",
  "success",
  "info",
  "warning",
  "danger",
  "light",
  "dark",
  "white",
  "black",
  "link",
);
export type Colors = (typeof COLORS)[number];

export const GREY_COLORS = tuple(
  "black-bis",
  "black-ter",
  "grey-darker",
  "grey-dark",
  "grey",
  "grey-light",
  "grey-lighter",
  "white-ter",
  "white-bis",
);
export type GreyColors = (typeof GREY_COLORS)[number];

export const TEXT_ALIGNMENTS = tuple("centered", "justified", "left", "right");
export type TextAlignments = (typeof TEXT_ALIGNMENTS)[number];

export const TEXT_SIZES = tuple(1, 2, 3, 4, 5, 6);
export type TextSizes = (typeof TEXT_SIZES)[number];

export const TEXT_TRANSFORMS = tuple("capitalized", "lowercase", "uppercase");
export type TextTransforms = (typeof TEXT_TRANSFORMS)[number];

export const TEXT_WEIGHTS = tuple("light", "normal", "semibold", "bold");
export type TextWeights = (typeof TEXT_WEIGHTS)[number];

export type TypographyHelpersProps = Partial<{
  backgroundColor: Colors | GreyColors;
  italic: boolean;
  textAlignment: TextAlignments;
  textColor: Colors | GreyColors;
  textSize: TextSizes;
  textTransform: TextTransforms;
  textWeight: TextWeights;
}>;

export const typographyHelpersPropTypes = {
  backgroundColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  italic: PropTypes.bool,
  textAlignment: PropTypes.oneOf(TEXT_ALIGNMENTS),
  textColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  textSize: PropTypes.oneOf(TEXT_SIZES),
  textTransform: PropTypes.oneOf(TEXT_TRANSFORMS),
  textWeight: PropTypes.oneOf(TEXT_WEIGHTS),
};

export const transformTypographyHelpers: TransformFunc<
  TypographyHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    typographyHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    backgroundColor,
    className,
    italic,
    textAlignment,
    textColor,
    textSize,
    textTransform,
    textWeight,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        [`has-background-${backgroundColor}`]: backgroundColor,
        [`has-text-${textColor}`]: textColor,
        "is-italic": italic,
        [`is-${textTransform}`]: textTransform,
        [`has-text-${textAlignment}`]: textAlignment,
        [`has-text-weight-${textWeight}`]: textWeight,
        [`is-size-${textSize}`]: textSize,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Visibility
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const DISPLAYS = tuple(
  "block",
  "flex",
  "inline",
  "inline-block",
  "inline-flex",
);
export type Displays = (typeof DISPLAYS)[number];

export type VisibilityHelpersProps = Partial<{
  /** Hides an element (unclear on where this is documented in Bulma) */
  hidden: boolean;
  /** Adds visibility hidden */
  invisible: boolean;
  /**
   * Hide elements visually but keep the element available to be announced by a
   * screen reader
   */
  srOnly: boolean;
}>;

export const visibilityHelpersPropTypes = {
  hidden: PropTypes.bool,
  invisible: PropTypes.bool,
  srOnly: PropTypes.bool,
};

export const transformVisibilityHelpers: TransformFunc<
  VisibilityHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    visibilityHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, hidden, invisible, srOnly, ...rest } = props;

  return {
    className: classNames(
      {
        "is-hidden": hidden,
        "is-invisible": invisible,
        "is-sr-only": srOnly,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Other
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OtherHelpersProps = Partial<{
  /** Removes any margin */
  marginless: boolean;
  /** Removes any padding */
  paddingless: boolean;
  /** Removes any radius */
  radiusless: boolean;
  /** Removes any shadow */
  shadowless: boolean;
  /** Prevents the text from being selectable */
  unselectable: boolean;
}>;

export const otherHelpersPropTypes = {
  marginless: PropTypes.bool,
  paddingless: PropTypes.bool,
  radiusless: PropTypes.bool,
  shadowless: PropTypes.bool,
  unselectable: PropTypes.bool,
};

export const transformOtherHelpers: TransformFunc<OtherHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    otherHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    className,
    marginless,
    paddingless,
    radiusless,
    shadowless,
    unselectable,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        "is-marginless": marginless,
        "is-paddingless": paddingless,
        "is-radiusless": radiusless,
        "is-shadowless": shadowless,
        "is-unselectable": unselectable,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Responsive
 */
export const BREAKPOINTS = tuple(
  "mobile",
  "tablet",
  "desktop",
  "widescreen",
  "fullhd",
  "touch",
);
export type Breakpoints = (typeof BREAKPOINTS)[number];

export type LimitiedResponsiveBreakpointProps = Partial<{
  display: {
    value: Displays;
  };
  hide: {
    value: boolean;
  };
  textAlignment: {
    value: TextAlignments;
  };
  textSize: {
    value: TextSizes;
  };
}>;

export type ResponsiveBreakpointProps = Partial<{
  display: {
    only?: boolean;
    value: Displays;
  };
  hide: {
    only?: boolean;
    value: boolean;
  };
  textAlignment: {
    only?: boolean;
    value: TextAlignments;
  };
  textSize: {
    only?: boolean;
    value: TextSizes;
  };
}>;

export type ResponsiveHelpersProps = Partial<{
  responsive: Partial<{
    mobile: LimitiedResponsiveBreakpointProps;
    tablet: ResponsiveBreakpointProps;
    desktop: ResponsiveBreakpointProps;
    widescreen: ResponsiveBreakpointProps;
    fullhd: LimitiedResponsiveBreakpointProps;
    touch: LimitiedResponsiveBreakpointProps;
  }>;
}>;

export const responsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const limitedResponsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const responsiveHelpersPropTypes = {
  responsive: PropTypes.shape({
    // tslint:disable:object-literal-sort-keys
    mobile: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    tablet: PropTypes.shape(responsiveBreakpointPropTypes),
    desktop: PropTypes.shape(responsiveBreakpointPropTypes),
    widescreen: PropTypes.shape(responsiveBreakpointPropTypes),
    fullhd: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    touch: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    // tslint:enable:object-literal-sort-keys
  }),
};

export const transformResponsiveHelpers: TransformFunc<
  ResponsiveHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    responsiveHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, responsive, ...rest } = props;

  return {
    className: classNames(
      responsive !== undefined
        ? Object.keys(responsive)
            .filter(breakpoint => responsive[breakpoint] !== undefined)
            .map(breakpoint => {
              const names = {};
              const { display, hide, textAlignment, textSize } = responsive[
                breakpoint
              ] as
                | ResponsiveBreakpointProps
                | LimitiedResponsiveBreakpointProps;
              if (display !== undefined) {
                const value = display.value;
                const only = "only" in display ? display.only === true : false;
                names[
                  `is-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (hide !== undefined) {
                const value = hide.value;
                const only = "only" in hide ? hide.only === true : false;
                names[`is-hidden-${breakpoint}${only ? "-only" : ""}`] = value;
              }
              if (textAlignment !== undefined) {
                const value = textAlignment.value;
                const only =
                  "only" in textAlignment ? textAlignment.only === true : false;
                names[
                  `has-text-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (textSize !== undefined) {
                const value = textSize.value;
                const only =
                  "only" in textSize ? textSize.only === true : false;
                names[
                  `is-size-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }

              return names;
            })
            .reduce((acc, cv) => ({ ...acc, ...cv }), {})
        : undefined,
      className,
    ),
    ...rest,
  };
};

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const helpersPropTypes = {
  ...floatHelpersPropTypes,
  ...overflowHelpersPropTypes,
  ...overlayHelpersPropTypes,
  ...typographyHelpersPropTypes,
  ...visibilityHelpersPropTypes,
  ...otherHelpersPropTypes,
  ...responsiveHelpersPropTypes,
  className: PropTypes.string,
};

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);

export type TransformFunc<TTransformProps extends {}> = <
  TProps extends TTransformProps & { className?: string }
>(
  props: TProps,
  componentName: string,
  location?: string,
) => Omit<TProps, keyof TTransformProps | "className"> & { className?: string };
