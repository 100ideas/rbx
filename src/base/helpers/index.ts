import { Prefer } from "../../types";
import { TransformFunc } from "./types";

import { FloatHelpersProps, transformFloatHelpers } from "./float";
import { OtherHelpersProps, transformOtherHelpers } from "./other";
import { OverflowHelpersProps, transformOverflowHelpers } from "./overflow";
import { OverlayHelpersProps, transformOverlayHelpers } from "./overlay";
import {
  ResponsiveHelpersProps,
  transformResponsiveHelpers,
} from "./responsive";
import {
  transformTypographyHelpers,
  TypographyHelpersProps,
} from "./typography";
import {
  transformVisibilityHelpers,
  VisibilityHelpersProps,
} from "./visibility";

export { TransformFunc } from "./types";

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);
