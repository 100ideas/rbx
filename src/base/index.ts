export { forwardRefAs } from "./exotic";
export { Generic, propTypes as genericPropTypes } from "./generic";
export { HelpersProps, makeRootValidatingTransform } from "./helpers";
