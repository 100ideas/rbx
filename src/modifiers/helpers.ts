import { cx } from "emotion";

import { makeTransform } from "./utils";

export type HelpersProps = Partial<{
  clearfix: boolean;
  clipped: boolean;
  hidden: boolean;
  invisible: boolean;
  marginless: boolean;
  overlay: boolean;
  paddingless: boolean;
  pull: "right" | "left";
  radiusless: boolean;
  shadowless: boolean;
  unselectable: boolean;
}>;

export const transformHelpersModifiers = makeTransform<HelpersProps>(
  props =>
    cx(props.className, {
      "is-clearfix": props.clearfix,
      "is-clipped": props.clipped,
      "is-hidden": props.hidden,
      "is-invisible": props.invisible,
      "is-marginless": props.marginless,
      "is-overlay": props.overlay,
      "is-paddingless": props.paddingless,
      [`is-pulled-${props.pull}`]: props.pull,
      "is-radiusless": props.radiusless,
      "is-shadowless": props.shadowless,
      "is-unselectable": props.unselectable,
    }),
  [
    "clearfix",
    "clipped",
    "hidden",
    "invisible",
    "marginless",
    "overlay",
    "paddingless",
    "pull",
    "radiusless",
    "shadowless",
    "unselectable",
  ],
);
