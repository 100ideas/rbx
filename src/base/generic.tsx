import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs } from "./exotic";
import { HelpersProps } from "./helpers";
import { TransformContext } from "./transform-context";

export type GenericProps = HelpersProps;

export const propTypes = {
  /** The component this should render as */
  as: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.string,
    PropTypes.shape({ render: PropTypes.func.isRequired }),
  ]),
};

export const Generic = Object.assign(
  forwardRefAs<GenericProps, "div">(
    (props, ref) => {
      const { as, ...rest } = props;
      return (
        <TransformContext.Consumer>
          {({ transform }) => {
            const transformed = transform(rest, "Generic");
            return React.createElement(as!, { ref, ...transformed });
          }}
        </TransformContext.Consumer>
      );
    },
    { as: "div" },
  ),
  { propTypes },
);
