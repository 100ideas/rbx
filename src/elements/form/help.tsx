import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";
import { Variables } from "../../base/helpers/variables";

export type HelpModifierProps = Partial<{
  color: Variables["colors"];
}>;

export type HelpOwnProps = HelpersProps & HelpModifierProps;
export type HelpForwardsProps = { className: string };

export const Help = forwardRefAs<"p", HelpOwnProps, HelpForwardsProps>(
  ({ className, color, ...rest }, ref) => (
    <Generic
      className={classNames("help", { [`is-${color}`]: color }, className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "p" },
);

Help.displayName = "Help";
Help.propTypes = {
  color: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};
