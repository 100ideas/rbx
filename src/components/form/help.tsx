import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/components/exotic";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { Colors } from "@/modifiers/color";

export type HelpModifierProps = Partial<{
  color: Colors;
}>;

export type HelpProps = ModifierProps & HelpModifierProps;

export const Help = forwardRefAs<HelpProps, "p">((props, ref) => {
  const { as, color, ...rest } = transformModifiers(props);
  rest.className = cx("help", rest.className, {
    [`is-${color}`]: color,
  });
  return React.createElement(as!, { ref, ...rest });
}, "p");
