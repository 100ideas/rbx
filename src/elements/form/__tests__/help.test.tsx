import { COLORS } from "../../../base/helpers";
import { Help } from "../help";

import {
  hasProperties,
  makeNodeFactory,
  makeShallowWrapper,
  testForwardRefAsExoticComponentIntegration,
  testThemeIntegration,
  validateOneOfPropType,
} from "../../../__tests__/testing";

const COMPONENT = Help;
const COMPONENT_NAME = "Help";
const DEFAULT_ELEMENT = "p";
const BULMA_CLASS_NAME = "help";

const makeNode = makeNodeFactory(COMPONENT);

describe(`${COMPONENT_NAME} component`, () => {
  hasProperties(COMPONENT, {
    defaultProps: { as: DEFAULT_ELEMENT },
  });

  testForwardRefAsExoticComponentIntegration(
    makeNode,
    makeShallowWrapper,
    DEFAULT_ELEMENT,
    BULMA_CLASS_NAME,
  );

  testThemeIntegration(makeNode, makeShallowWrapper);

  describe("props", () => {
    const { propTypes } = COMPONENT;

    describe("color", () => {
      validateOneOfPropType(propTypes, "color", COLORS);

      COLORS.map(color =>
        it(`should be ${color}`, () => {
          const node = makeNode({ color });
          const wrapper = makeShallowWrapper(node);
          expect(wrapper.hasClass(`is-${color}`)).toBe(true);
        }),
      );
    });
  });
});
