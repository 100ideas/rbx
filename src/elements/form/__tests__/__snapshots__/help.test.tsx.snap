// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help component should be displayed as a successful message 1`] = `
<p
  className="help is-success"
>
  <p>
    Default
  </p>
</p>
`;

exports[`Help component should concat classname in props with Bulma classname 1`] = `
<p
  className="help other-class this-is-a-test"
>
  <p>
    Default
  </p>
</p>
`;

exports[`Help component should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "as": "p",
  },
  "render": [Function],
}
`;

exports[`Help component should have help classname 1`] = `
<p
  className="help"
>
  <p
    className="bd-notification is-success"
  >
    <p>
      Default
    </p>
    <p>
      Container
    </p>
  </p>
</p>
`;

exports[`Help component should use inline styles 1`] = `
<p
  className="help"
  style={
    Object {
      "height": 250,
    }
  }
>
  <p>
    Default
  </p>
</p>
`;
